sw/device/tests/i2c_host_clock_stretching_test.c

```diff
@@ -0,0 +1,255 @@
+            "IRQ from incorrect peripheral: exp = %d(i2c0), found = %d",
+            kTopEarlgreyPlicPeripheralI2c0, peripheral);
+
+  irq_fired =
+      (dif_i2c_irq_t)(plic_irq_id - (dif_rv_plic_irq_id_t)
+                                        kTopEarlgreyPlicIrqIdI2c0FmtThreshold);
+
+  LOG_INFO("%s: plic:%d, i2c:%d", __func__, plic_irq_id, irq_fired);
+  TRY(dif_i2c_irq_acknowledge(&i2c, irq_fired));
+
+  // Complete the IRQ at PLIC.
+  TRY(dif_rv_plic_irq_complete(&plic, kHart, plic_irq_id));
+  return OK_STATUS();
+}
+
+void ottf_external_isr(void) { test_result = external_isr(); }
+
+// We make this a macro to show the caller line number when printing a error.
+#define CHECK_IRQ_EQ(irq)                                                \
+  do {                                                                   \
+    irq_global_ctrl(false);                                              \
+    if (irq_fired == UINT32_MAX) {                                       \
+      wait_for_interrupt();                                              \
+      irq_global_ctrl(true);                                             \
+    }                                                                    \
+    TRY_CHECK(irq_fired == irq, "Got %d, expected %d.", irq_fired, irq); \
+  } while (false)
+
+static status_t sw_reset() {
+  uint8_t write_buffer[2] = {kSwResetReg, 0x01};
+  TRY(i2c_testutils_write(&i2c, kDeviceAddr, sizeof(write_buffer), write_buffer,
+                          true));
+  return OK_STATUS();
+}
+static status_t read_manufacture_id(void) {
+  uint8_t reg = kManufacturerIdReg, data = 0;
+  TRY(sw_reset());
+  TRY(i2c_testutils_write(&i2c, kDeviceAddr, 1, &reg, true));
+  TRY(i2c_testutils_read(&i2c, kDeviceAddr, 1, &data));
+  TRY_CHECK(data == kManufacturerId, "Unexpected value %x", data);
+  return OK_STATUS();
+}
+
+static status_t read_product_id(void) {
+  uint8_t reg = kProductIdReg, data = 0;
+  TRY(i2c_testutils_write(&i2c, kDeviceAddr, 1, &reg, true));
+  TRY(i2c_testutils_read(&i2c, kDeviceAddr, 1, &data));
+  TRY_CHECK(data == kProductId, "Unexpected value %x", data);
+  return OK_STATUS();
+}
+
+static status_t rx_stretch_timeout(void) {
+  enum { kTimeoutMillis = 15 };
+  TRY(dif_i2c_irq_set_enabled(&i2c, kDifI2cIrqStretchTimeout,
+                              kDifToggleEnabled));
+
+  // Set the delay that will make the target to stretch the clock.
+  {
+    uint8_t write_buffer[2] = {kRxDelayReg, kTimeoutMillis};
+    TRY(i2c_testutils_write(&i2c, kDeviceAddr, sizeof(write_buffer),
+                            write_buffer, true));
+  }
+
+  uint32_t cycles = (kTimeoutMillis - 1) * 100;
+  TRY(dif_i2c_enable_clock_stretching_timeout(&i2c, kDifToggleEnabled, cycles));
+
+  uint8_t reg = kProductIdReg;
+  uint8_t data = 0;
+  TRY(i2c_testutils_write(&i2c, kDeviceAddr, 1, &reg, true));
+
+  irq_fired = UINT32_MAX;
+  TRY(i2c_testutils_issue_read(&i2c, kDeviceAddr, 1));
+
+  irq_fired = UINT32_MAX;
```

Comment:
If you're expecting a clock-stretch event to be triggered by the previous read, I don't think you want to set `irq_fired` here: you should set it before issuing the read.

To be safe, I'd wrap assignments to this value in non-irq contexts with a disable/enable sequence (even when assigning before issuing the read):
```
irq_global_ctrl(false);
irq_fired = <value>;
irq_global_ctrl(true);
```

Reply:
Fixed.